### PR TITLE
EOS-16741: Dynamic registration and deregistration of network transports

### PR DESCRIPTION
**Pull request by papan-singh**

### CODE REVIEW

File: net/test/ut/client_server.c [hunk 1]
```diff
@@ -401,6 +402,20 @@ void m0_net_test_client_server_bulk_ut(void)
...
+	print_xprt();
+	m0_mem_xprt_fini();
+	m0_mem_xprt_init();
+	print_xprt();
+
+	m0_net_lnet_fini();
+	m0_net_lnet_init();
+	print_xprt();
+
+//m0_net_sock_mod_init();
```
Comment (upendrapatwardhan) [inline net/test/ut/client_server.c:416]: Remove unwanted code

File: motr/motr-pub.api [hunk 2]
```diff
@@ -460,3 +460,4 @@ m0_iem
 m0_console_printf
 m0_console_flush
 m0_error_printf
+print_xprt
```
Comment (upendrapatwardhan) [inline motr/motr-pub.api:463]: m0_net_xprt_printf?

**Review by upendrapatwardhan — APPROVED** [review]

File: net/net.c [hunk 3]
```diff
@@ -196,18 +196,26 @@ M0_EXPORTED(m0_net_xprt_register);
...
 		}
 	}
 	M0_IMPOSSIBLE("Wrong xprt.");
 }
 M0_EXPORTED(m0_net_xprt_deregister);
+
+void print_xprt(void)
+{
+	int i;
+	for (i = 0; i < ARRAY_SIZE(xprts) -1; ++i)
```
Comment (huanghua78) [inline net/net.c:216]: Why is the upper limit "ARRAY_SIZE(xprts) -1"?
Should it be:
```
for (i = 0; i < ARRAY_SIZE(xprts); ++i)
```

And another bug:
"xprts[i]" might be NULL.
It should be check before dereferencing its ->nx_name.

File: motr/motr-pub.api (hunk 2 above)
Comment (huanghua78) [reply motr/motr-pub.api:463]: This is wrong.

Public API means, this API is callable by Motr application, linking against libmotr.so.
There is not need to export this as an public API.

File: net/test/ut/client_server.c [hunk 4]
```diff
@@ -401,6 +402,20 @@ void m0_net_test_client_server_bulk_ut(void)
 			       64, 0x100000,
 			       8, 16, 0x1000, 0x10000);
 }
+void m0_net_test_xprt_dymanic_reg_dereg_ut(void)
+{
+	print_xprt();
+	m0_mem_xprt_fini();
+	m0_mem_xprt_init();
+	print_xprt();
```
Comment (huanghua78) [inline net/test/ut/client_server.c:410]: print_xprt() will print something. But it does not check anything.
Developers  must look into the tracing logs and manually make sure the result is success.

This is wrong.

Here in UT, we need to check that:
after m0_mem_xprt_fini(), the "m0_net_bulk_mem_xprt" is not in the global array.
If it is default, the deault_xprt should be NULL.

After m0_mem_xprt_init(), the "m0_net_bulk_mem_xprt" is again in the global array.

File: motr/motr-pub.api (hunk 2 above)
Comment (papan-singh) [reply motr/motr-pub.api:463]: incorporated

File: net/net.c (hunk 3 above)
Comment (papan-singh) [reply net/net.c:216]: incorporated

File: net/test/ut/client_server.c (hunk 4 above)
Comment (papan-singh) [reply net/test/ut/client_server.c:410]: incorporated

File: net/test/ut/client_server.c (hunk 1 above)
Comment (papan-singh) [reply net/test/ut/client_server.c:416]: incorprated

File: net/net.c [hunk 5]
```diff
@@ -196,18 +196,49 @@ M0_EXPORTED(m0_net_xprt_register);
...
+{
+	int i;
+	for (i = 0; i < ARRAY_SIZE(xprts); ++i) {
+		if (xprts[i] != NULL)
+			M0_LOG(M0_ALWAYS, "xprt name:%s", xprts[i]->nx_name);
+	}
+}
+M0_EXPORTED(m0_net_print_xprt);
+
+M0_INTERNAL int m0_net_check_xprt(const struct m0_net_xprt *xprt)
```
[the remark this answers is not in the document]
Comment (papan-singh) [reply net/net.c:223]: incorporated

File: net/net.c [hunk 6]
```diff
@@ -196,18 +196,49 @@ M0_EXPORTED(m0_net_xprt_register);
...
+			for (j = i; j < ARRAY_SIZE(xprts) - 1; ++j)
+				xprts[j] = xprts[j + 1];
 			return;
 		}
 	}
 	M0_IMPOSSIBLE("Wrong xprt.");
 }
 M0_EXPORTED(m0_net_xprt_deregister);
+
+M0_INTERNAL void m0_net_print_xprt(void)
```
[the remark this answers is not in the document]
Comment (papan-singh) [reply net/net.c:213]: incorporated

File: net/test/ut/client_server.c [hunk 7]
```diff
@@ -401,6 +402,31 @@ void m0_net_test_client_server_bulk_ut(void)
 			       64, 0x100000,
 			       8, 16, 0x1000, 0x10000);
 }
+void m0_net_test_xprt_dymanic_reg_dereg_ut(void)
+{
+	M0_LOG(M0_ALWAYS, "Before mem fini\n");
+	m0_net_print_xprt();
```
[the remark this answers is not in the document]
Comment (papan-singh) [reply net/test/ut/client_server.c:408]: incorporated

File: net/net.c [hunk 8]
```diff
@@ -196,18 +196,49 @@ M0_EXPORTED(m0_net_xprt_register);
...
+
+	for (i = 0; i < ARRAY_SIZE(xprts); ++i)
+	{
+		if (xprts[i] == xprt)
+			found = true;
+	}
+
+	d_xprt = m0_net_xprt_default_get();
+	if ((found == false) && (d_xprt == xprt))
+		xprt_default = NULL;
```
Comment (huanghua78) [inline net/net.c:237]: No need for this.
As I said, this is checking function. It should not modify the "xprt_default" variable.

If this happens, it means something wrong in the deregister function. It should not use this function to correct that.

Also:
we use
```
if ((!found) && (d_xprt == xprt))
		xprt_default = NULL;
```
according to coding style.

Comment (papan-singh) [reply net/net.c]: incorporated

File: net/net.c [hunk 9]
```diff
@@ -196,18 +196,49 @@ M0_EXPORTED(m0_net_xprt_register);
...
 	int i;
-
+	int j;
 	for (i = 0; i < ARRAY_SIZE(xprts); ++i) {
 		if (xprts[i] == xprt) {
-			xprts[i] = NULL;
 			if (xprt == xprt_default)
 				xprt_default = NULL;
+			for (j = i; j < ARRAY_SIZE(xprts) - 1; ++j)
+				xprts[j] = xprts[j + 1];
```
[the remark this answers is not in the document]
Comment (papan-singh) [reply net/net.c:205]: incorporated

**Review by huanghua78 — APPROVED** [review]

**Comment by cortx-admin** [comment]
<h1>Jenkins CI Result : <a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259/>Motr#259</a></h1><h2>Motr Test Summary</h2><table><tr><th>Test Result</th><th>Count</th><th>Info</tr><tr><td>:x:Failed</td><td>2</td><td><details><summary>:file_folder:</summary><p><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/00userspace-tests target=_blank>01motr-single-node/00userspace-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/28motr-sys-kvs-test target=_blank>01motr-single-node/28motr-sys-kvs-test</a></p></details></td></tr><tr><td>:checkered_flag:Skipped</td><td>5</td><td><details><summary>:file_folder:</summary><p><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/37protocol target=_blank>01motr-single-node/37protocol</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/04motr-single-node/31motr-sys-kvs-kernel-test target=_blank>04motr-single-node/31motr-sys-kvs-kernel-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/07multi-client-tests target=_blank>05motr-single-node/07multi-client-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/39mount-fail target=_blank>05motr-single-node/39mount-fail</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/12sns-cc-io-tests target=_blank>05motr-single-node/12sns-cc-io-tests</a></p></details></td></tr><tr><td>:heavy_check_mark:Passed</td><td>57</td><td><details><summary>:file_folder:</summary><p><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/46m0crate target=_blank>01motr-single-node/46m0crate</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/01kernel-tests target=_blank>01motr-single-node/01kernel-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/03initscript-tests target=_blank>01motr-single-node/03initscript-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/30hagen-test target=_blank>01motr-single-node/30hagen-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/29confgen-test target=_blank>01motr-single-node/29confgen-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/35m0singlenode target=_blank>01motr-single-node/35m0singlenode</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/04net-tests target=_blank>01motr-single-node/04net-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/05rpcping-tests target=_blank>01motr-single-node/05rpcping-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/06console-tests target=_blank>01motr-single-node/06console-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/01motr-single-node/52singlenode-sanity target=_blank>01motr-single-node/52singlenode-sanity</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/02motr-single-node/20pool-version-test target=_blank>02motr-single-node/20pool-version-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/02motr-single-node/21spiel-test target=_blank>02motr-single-node/21spiel-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/02motr-single-node/47motr-tests-user-kernel target=_blank>02motr-single-node/47motr-tests-user-kernel</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/02motr-single-node/09poolmach-tests target=_blank>02motr-single-node/09poolmach-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/02motr-single-node/41motr-conf-update target=_blank>02motr-single-node/41motr-conf-update</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/02motr-single-node/08degraded-mode-tests target=_blank>02motr-single-node/08degraded-mode-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/02motr-single-node/24m0d-fsync-test target=_blank>02motr-single-node/24m0d-fsync-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/02motr-single-node/02system-tests target=_blank>02motr-single-node/02system-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/02motr-single-node/22rpc-cancel-test target=_blank>02motr-single-node/22rpc-cancel-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/02motr-single-node/53clusterusage-alert target=_blank>02motr-single-node/53clusterusage-alert</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/02motr-single-node/23m0d-signal-test target=_blank>02motr-single-node/23m0d-signal-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/02motr-single-node/51kem target=_blank>02motr-single-node/51kem</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/03motr-single-node/54sns-repair-motr-1f target=_blank>03motr-single-node/54sns-repair-motr-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/03motr-single-node/59sns-repair-motr-abort target=_blank>03motr-single-node/59sns-repair-motr-abort</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/03motr-single-node/17sss-st-test target=_blank>03motr-single-node/17sss-st-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/03motr-single-node/58sns-repair-motr-ios-fail target=_blank>03motr-single-node/58sns-repair-motr-ios-fail</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/03motr-single-node/60sns-repair-motr-abort-quiesce target=_blank>03motr-single-node/60sns-repair-motr-abort-quiesce</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/03motr-single-node/55sns-repair-motr-1n-1f target=_blank>03motr-single-node/55sns-repair-motr-1n-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/03motr-single-node/56sns-repair-motr-mf target=_blank>03motr-single-node/56sns-repair-motr-mf</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/03motr-single-node/61spiel-multi-confd target=_blank>03motr-single-node/61spiel-multi-confd</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/03motr-single-node/16conf-st-test target=_blank>03motr-single-node/16conf-st-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/03motr-single-node/57sns-repair-motr-1k-1f target=_blank>03motr-single-node/57sns-repair-motr-1k-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/03motr-single-node/36spare-reservation target=_blank>03motr-single-node/36spare-reservation</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/04motr-single-node/51motr-rmw target=_blank>04motr-single-node/51motr-rmw</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/04motr-single-node/50motr-rm-lock-cc-io target=_blank>04motr-single-node/50motr-rm-lock-cc-io</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/04motr-single-node/25spiel-sns-repair-test target=_blank>04motr-single-node/25spiel-sns-repair-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/04motr-single-node/33motr-st target=_blank>04motr-single-node/33motr-st</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/04motr-single-node/19sns-abort-test target=_blank>04motr-single-node/19sns-abort-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/04motr-single-node/32m0t1fs-rconfc-fail-test target=_blank>04motr-single-node/32m0t1fs-rconfc-fail-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/04motr-single-node/34sns-repair-1n-1f target=_blank>04motr-single-node/34sns-repair-1n-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/04motr-single-node/26spiel-sns-repair-quiesce-test target=_blank>04motr-single-node/26spiel-sns-repair-quiesce-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/04motr-single-node/49motr-rpc-cancel target=_blank>04motr-single-node/49motr-rpc-cancel</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/04motr-single-node/27sns-repair-io-fail-test target=_blank>04motr-single-node/27sns-repair-io-fail-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/04motr-single-node/48motr-raid0-io target=_blank>04motr-single-node/48motr-raid0-io</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/40motr-dgmode target=_blank>05motr-single-node/40motr-dgmode</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/42motr-utils target=_blank>05motr-single-node/42motr-utils</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/13sns-repair-quiesce target=_blank>05motr-single-node/13sns-repair-quiesce</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/45motr-sns-repair-N-1 target=_blank>05motr-single-node/45motr-sns-repair-N-1</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/11sns-multi-tests target=_blank>05motr-single-node/11sns-multi-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/44motr-sns-repair target=_blank>05motr-single-node/44motr-sns-repair</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/14m0t1fs-fsync-test target=_blank>05motr-single-node/14m0t1fs-fsync-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/43motr-sync-replication target=_blank>05motr-single-node/43motr-sync-replication</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/38sns-abort-quiesce-test target=_blank>05motr-single-node/38sns-abort-quiesce-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/15m0t1fs-fwait-test target=_blank>05motr-single-node/15m0t1fs-fwait-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/18m0mount-test target=_blank>05motr-single-node/18m0mount-test</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/05motr-single-node/10sns-single-tests target=_blank>05motr-single-node/10sns-single-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit/motr-single-node/62dix-repair-st-lookup-insert target=_blank>motr-single-node/62dix-repair-st-lookup-insert</a></p></details></td></tr><tr><td>Total</td><td>64</td><td><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/259//testReport/junit>:link:</a></td></tr></table><h2> CppCheck Summary</h2><p> &emsp;&emsp;&emsp;Cppcheck: No new warnings found :thumbsup:</p>

**Pull request merged**